onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp

```diff
@@ -617,18 +688,94 @@
         return outputShape;
     }
 
+    void BroadcastTensorShapeAndSetStrides(
+        std::vector<DimensionType>& inputShape0,
+        std::vector<DimensionType>& inputStride0,
+        std::vector<DimensionType>& inputShape1,
+        std::vector<DimensionType>& inputStride1
+        )
+    {
+        if (inputShape0 != inputShape1)
+        {
+            auto outputRank = std::max(inputShape0.size(), inputShape1.size());
+
+            // Walk backwards through both input shapes and broadcast each dimension,
+            // ignoring the last 2 dimensions (matrix dimensions).
+            auto inDim0Iter = inputShape0.rbegin() + 2;
+            auto inDim1Iter = inputShape1.rbegin() + 2;
+
+            auto inStride0Iter = inputStride0.rbegin() + 2;
+            auto inStride1Iter = inputStride1.rbegin() + 2;
+            outputRank -= 2;
+
+            std::vector<DimensionType> newInputShape(outputRank);
+            std::vector<DimensionType> newInputStride0(outputRank), newInputStride1(outputRank);
+
+            while (outputRank-- > 0)
+            {
+                DimensionType inDimension0 = 1;
+                DimensionType inStride0 = 0;
+                if (inDim0Iter != inputShape0.rend())
+                {
+                    inDimension0 = *inDim0Iter;
+                    inStride0 = *inStride0Iter;
+                    ++inDim0Iter;
+                    ++inStride0Iter;
+                }
+
+                DimensionType inDimension1 = 1;
+                DimensionType inStride1 = 0;
+                if (inDim1Iter != inputShape1.rend())
+                {
+                    inDimension1 = *inDim1Iter;
+                    inStride1 = *inStride1Iter;
+                    ++inDim1Iter;
+                    ++inStride1Iter;
+                }
+
+                // 0-sized dimensions indicate an empty tensor and shouldn't be broadcasted to higher dimensions.
+                // TODO: 0-sized dimensions are actually scalars and are broadcastable. Can this be made to work with
+                // scalars?
```

Comment:
Sorry, this TODO is deletable. I thought it was talking about empty dimensions (a 0D scalar) as opposed to a 0 size in the middle of dimensions (e.g. [2,0,3]).